File: scala/csum/SimpleSummer.scala
class SimpleSummer extends Checksummer {
	var byteCount = 0L
	var byteSum   = 0L

	def start() : Unit = {
		byteCount = 0L
		byteSum   = 0L
	}

	def accumulate(chs: Array[Byte], len: Int) : Unit = {
		byteCount += len
		for (i <- 0 until len)
			byteSum += chs(i).toLong & 0xff
	}

	def finish() : Unit = {
		// Nothing for this summer
	}

	def getStringState() : String = {
		"%016x".format(byteCount) + "_" + "%016x".format(byteSum)
	}

	def getStringSum() : String = {
		"%016x".format(byteCount) + "_" + "%016x".format(byteSum)
	}
}

File: go/src/csums/SimpleSummer.go
package csums

import (
	"fmt"
)

type SimpleSummer struct {
	// extends Checksummer is not needed in go.  just impl the fncs.
	byteCount uint64
	byteSum   uint64
}

func (simpleSummer *SimpleSummer) Start() {
	simpleSummer.byteCount = 0
	simpleSummer.byteSum = 0
}

func (simpleSummer *SimpleSummer) Accumulate(bytes []byte, n int) {
	simpleSummer.byteCount += uint64(n)
	for i := 0; i < n; i++ {
		simpleSummer.byteSum += uint64(bytes[i])
	}
}

func (simpleSummer *SimpleSummer) Finish() {
}

func (simpleSummer SimpleSummer) GetStringState() (state string) {
	return fmt.Sprintf("%016x_%016x", simpleSummer.byteCount, simpleSummer.byteSum)
}

func (simpleSummer SimpleSummer) GetStringSum() (sum string) {
	return fmt.Sprintf("%016x_%016x", simpleSummer.byteCount, simpleSummer.byteSum)
}
